coordinator/src/main/java/org/apache/uniffle/coordinator/AppBalanceSelectStorageStrategy.java

```diff
@@ -39,106 +45,127 @@
-            .sorted(Comparator.comparingInt(entry -> entry.getValue().getAppNum().get())).collect(Collectors.toList());
-
-    for (Map.Entry<String, RankValue> entry : sizeList) {
-      String storagePath = entry.getKey();
-      if (availableRemoteStorageInfo.containsKey(storagePath)) {
-        appIdToRemoteStorageInfo.putIfAbsent(appId, availableRemoteStorageInfo.get(storagePath));
-        incRemoteStorageCounter(storagePath);
-        break;
-      }
-    }
-    return appIdToRemoteStorageInfo.get(appId);
-  }
-
-  @Override
   @VisibleForTesting
-  public synchronized void incRemoteStorageCounter(String remoteStoragePath) {
-    RankValue counter = remoteStoragePathCounter.get(remoteStoragePath);
-    if (counter != null) {
-      counter.getAppNum().incrementAndGet();
-    } else {
-      // it may be happened when assignment remote storage
-      // and refresh remote storage at the same time
-      LOG.warn("Remote storage path lost during assignment: {} doesn't exist, reset it to 1",
-          remoteStoragePath);
-      remoteStoragePathCounter.put(remoteStoragePath, new RankValue(1));
+  public List<Map.Entry<String, RankValue>> sortPathByRankValue(
+      String path, String test, boolean isHealthy) {
+    try {
+      FileSystem fs = HadoopFilesystemProvider.getFilesystem(new Path(path), hdfsConf);
+      fs.delete(new Path(test),true);
+      if (isHealthy) {
+        RankValue rankValue = remoteStoragePathRankValue.get(path);
+        remoteStoragePathRankValue.put(path, new RankValue(0, rankValue.getAppNum().get()));
+      }
+    } catch (Exception e) {
+      RankValue rankValue = remoteStoragePathRankValue.get(path);
+      remoteStoragePathRankValue.put(path, new RankValue(Long.MAX_VALUE, rankValue.getAppNum().get()));
+      LOG.error("Failed to sort, we will not use this remote path {}.", path, e);
     }
+    return Lists.newCopyOnWriteArrayList(remoteStoragePathRankValue.entrySet()).stream()
+        .filter(Objects::nonNull).collect(Collectors.toList());
   }
 
   @Override
-  @VisibleForTesting
-  public synchronized void decRemoteStorageCounter(String storagePath) {
-    if (!StringUtils.isEmpty(storagePath)) {
-      RankValue atomic = remoteStoragePathCounter.get(storagePath);
-      if (atomic != null) {
-        double count = atomic.getAppNum().decrementAndGet();
-        if (count < 0) {
-          LOG.warn("Unexpected counter for remote storage: {}, which is {}, reset to 0",
-              storagePath, count);
-          atomic.getAppNum().set(0);
+  public List<Map.Entry<String, RankValue>> detectStorage(String uri) {
+    if (uri.startsWith(ApplicationManager.REMOTE_PATH_SCHEMA.get(0))) {
+      setRemotePathIsHealthy(true);
+      Path remotePath = new Path(uri);
+      String rssTest = uri + "/rssTest";
+      Path testPath = new Path(rssTest);
+      try {
+        FileSystem fs = HadoopFilesystemProvider.getFilesystem(remotePath, hdfsConf);
+        for (int j = 0; j < readAndWriteTimes; j++) {
```

Comment:
If we just detect the health of storage, it seems unnecessary for us write and read them multiple times.

Reply:
For appBalance, it can indeed be changed once, but for io sampling, the number of times can be increased, perhaps we can set the parameter to 1 by default.

Reply:
For appBalance, it shouldn't be a configuration option.

Reply:
done.